optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/roster/RosterService.java

```diff
@@ -361,24 +371,64 @@ public void updateShiftsOfRoster(Roster newRoster) {
         }
     }
 
+    public void scheduleUpdateOfRoster(Roster newRoster) {
+        tenantIdToRosterUpdateFutureMap.compute(newRoster.getTenantId(),
+                (tenantId, task) -> {
+                    if (task != null && !task.isDone()) {
+                        tenantIdToNextRosterMap.put(tenantId, newRoster);
+                        return task;
+                    } else {
+                        // Remove any stale old best solution that were put in
+                        // JUST AFTER the tenantIdToNextRosterMap.containsKey(tenantId)
+                        // return false BUT BEFORE the task finished,
+                        // since it'll be a worse solution than this one
+                        tenantIdToNextRosterMap.remove(tenantId);
+                        return rosterUpdateExecutorService.submit(() -> {
+                            transactionTemplate.executeWithoutResult(ts -> updateShiftsOfRoster(newRoster));
+                            tenantIdToRosterUpdateFutureMap.remove(tenantId);
+                            if (tenantIdToNextRosterMap.containsKey(tenantId)) {
+                                scheduleUpdateOfRoster(tenantIdToNextRosterMap.remove(tenantId));
+                            }
+                        });
+                    }
+                });
+    }
+
     // ************************************************************************
     // Solver
     // ************************************************************************
 
     public void solveRoster(Integer tenantId) {
-        solverManager.solve(tenantId);
+        solverManager.solveAndListen(tenantId, this::buildRoster, this::scheduleUpdateOfRoster);
     }
 
     public void replanRoster(Integer tenantId) {
-        solverManager.replan(tenantId);
+        Roster roster = buildRoster(tenantId);
+        roster.setNondisruptivePlanning(true);
+        roster.setNondisruptiveReplanFrom(OffsetDateTime.now());
+
+        // Help Optaplanner by unassigning any shifts where the employee is unavailable
+        Map<String, ConstraintMatchTotal<HardMediumSoftLongScore>> constraintMatchTotalMap = scoreManager.explainScore(roster)
+                .getConstraintMatchTotalMap();
+        String CONSTRAINT_ID = ConstraintMatchTotal.composeConstraintId(IndictmentUtils.CONSTRAINT_MATCH_PACKAGE,
+                "Unavailable time slot for an employee");
+        constraintMatchTotalMap.get(CONSTRAINT_ID)
+                .getConstraintMatchSet()
+                .forEach(cm -> {
+                    Shift shift = (Shift) cm.getJustificationList().stream().filter(o -> o instanceof Shift).findAny().get();
```

Comment:
`Optional.get()` without `isPresent()` check. A simple way to avoid it is to use `forEach()` as the terminal operation and unassigning the employee there.

Reply:
Abbreviations as variable names should be avoided. So `constraintMatch` instead of `cm`.

Using a full name that has a meaning avoids the mental mapping overhead. When reading `cm.getJustificationList()` I have to go back and guess that elements of `getConstraintMatchSet()` are `ConstraintMatch`es. This is a simple example but it illustrates the mental mapping problem.

The Clean Code book from Robert C. Martin has a nice chapter on naming.